vowpalwabbit/distributionally_robust.h

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright (c) by respective owners including Yahoo!, Microsoft, and
+individual contributors. All rights reserved.  Released under a BSD
+license as described in the file LICENSE.
+ */
+
+#pragma once
+
+#include <algorithm>
+#include <limits>
+#include <stdexcept>
+
+namespace VW
+{
+namespace distributionally_robust
+{
+
+class ChiSquared
+{
+  public:
+    struct Duals
+    {
+      bool unbounded;
+      double kappa;
+      double gamma;
+      double beta;
+      double n;
+
+      double qfunc(double w, double r) { return unbounded ? 1 : -(gamma + (beta + r) * w) / ((n + 1) * kappa); }
+    };
+
+  private:
+    double alpha;
+    double tau;
+    double wmin;
+    double wmax;
+    double rmin;
+    double rmax;
+
+    double n;
+    double sumw;
+    double sumwsq;
+    double sumwr;
+    double sumwsqr;
+    double sumwsqrsq;
+
+    double delta;
+
+    bool duals_stale;
+    Duals duals;
+
+  public:
+    // alpha: confidence level
+    // tau: count decay time constant
+    explicit ChiSquared(double _alpha, 
+               double _tau, 
+               double _wmin = 0, 
+               double _wmax = std::numeric_limits<double>::infinity(),
+               double _rmin = 0,
+               double _rmax = 1)
+      : alpha(_alpha),
+        tau(_tau),
+        wmin(_wmin),
+        wmax(_wmax),
+        rmin(_rmin),
+        rmax(_rmax),
+        n(0),
+        sumw(0),
+        sumwsq(0),
+        sumwr(0),
+        sumwsqr(0),
+        sumwsqrsq(0),
+        delta(chisq_onedof_isf(alpha)),
+        duals_stale(true)
+      {
+        if (alpha > 1 || alpha <= 0)
+          {
+            throw std::invalid_argument("invalid alpha value");
```

Comment:
@rajan-chari what's your suggestion for avoiding throws here?

Reply:
Rather than std::invalid_argument, we should be using the THROW macro here